src/app/cli/src/init/transaction_snark_profiler.ml

```diff
@@ -512,63 +535,110 @@ let pending_coinbase_stack_target (t : Transaction.t) stack =
 let format_time_span ts =
   sprintf !"Total time was: %{Time.Span.to_string_hum}" ts
 
+let apply_transactions_and_keep_intermediate_ledgers
+    ~(txn_state_view : Zkapp_precondition.Protocol_state.View.t)
+    first_pass_ledger txns =
+  let first_pass_target_ledgers, partially_applied_txns =
+    List.fold_map txns ~init:first_pass_ledger ~f:(fun l txn ->
+        let l', txn' =
+          Transaction.forget txn
+          |> Sparse_ledger.apply_transaction_first_pass ~constraint_constants
+               ~global_slot:txn_state_view.global_slot_since_genesis
+               ~txn_state_view l
+          |> Or_error.ok_exn
+        in
+        (l', (l', txn')) )
+    |> snd |> List.unzip
+  in
+  let second_pass_target_ledgers, applied_txns =
+    let second_pass_ledger = List.last_exn first_pass_target_ledgers in
+    List.fold_map partially_applied_txns ~init:second_pass_ledger
+      ~f:(fun l txn ->
+        let l', txn' =
+          Sparse_ledger.apply_transaction_second_pass l txn |> Or_error.ok_exn
+        in
+        (l', (l', txn')) )
+    |> snd |> List.unzip
+  in
+  (first_pass_target_ledgers, second_pass_target_ledgers, applied_txns)
+
 (* This gives the "wall-clock time" to snarkify the given list of transactions, assuming
    unbounded parallelism. *)
 let profile_user_command (module T : Transaction_snark.S) sparse_ledger0
     (transitions : Transaction.Valid.t list) _ : string Async.Deferred.t =
-  let constraint_constants = Genesis_constants.Constraint_constants.compiled in
+  (* TODO: assert (Ledger_hash.equal (Sparse_ledger.merkle_root l) (Sparse_ledger.merkle_root final_sparse_ledger)) ; *)
   let txn_state_view = Lazy.force curr_state_view in
   let open Async.Deferred.Let_syntax in
+  let first_pass_target_ledgers, second_pass_target_ledgers, applied_txns =
+    apply_transactions_and_keep_intermediate_ledgers ~txn_state_view
+      sparse_ledger0 transitions
+  in
+  let final_first_pass_ledger = List.last_exn first_pass_target_ledgers in
+  List.iter second_pass_target_ledgers ~f:(fun l ->
+      assert (
+        Ledger_hash.equal
+          (Sparse_ledger.merkle_root l)
+          (Sparse_ledger.merkle_root final_first_pass_ledger) ) ) ;
```

Comment:
can the Merkle root of the final 1st-pass ledger be computed just once, before the `iter` loop?